Experimental bitstream compression support

LAK132 opened the pull request:
This PR adds an option to compress bitstreams. Tested with an XC7A200T(FBG484-2) (MEGA65 R3). 

This is needed for using open tooling with the MEGA65 as the uncompressed bitstreams are too large to fit in the core slots.

With the cores I tested this with, the compression step didn't seem to have any noticeable impact on build times.

LAK132 committed:
Add bitstream compression option

Signed-off-by: Lakira Ashley <lakridesagain@gmail.com>

LAK132 commented on CMakeLists.txt:
```diff
@@ -36,8 +38,6 @@ target_include_directories(yaml-cpp PUBLIC
 	)
 
-# Set the CXX standard and compile time for our code only.
-set(CMAKE_CXX_STANDARD 14)
```
It refused to build on my system with the standard being set here

mithro commented:
@kgugala - Could you take a look?

kgugala commented:
@LAK132 thanks for the PR, It seems the CI is red. There is a lot compilation errors like:

```
03:00:22 | /root/prjxray/prjxray/lib/xilinx/tests/xc7series/configuration_test.cc:281:24: error: 'struct prjxray::xilinx::Configuration<prjxray::xilinx::Series7>::PacketData' has no member named 'size'
03:00:22 |   281 |  EXPECT_EQ(packet_data.size(), 15 * 101);
```

You mentioned that you had to change C++ standard. @LAK132 which standard did you use in your local build?

LAK132 commented:
I just moved the standard setting thing to the top of the script, it's still C++14. one of the dependencies didn't like that the standard wasn't being set before it was compiled, so I moved it. 

I'll take a look at those errors, must be coming from something I didn't try to build

LAK132 committed:
Fix up configuration tests

Signed-off-by: Lakira Ashley <lakridesagain@gmail.com>

kgugala commented:
looks like all the databases (artix, zynq, kintex and spartan) builds failed with:

```
ValueError: invalid literal for int() with base 16: '<0'
```

in the `005-tilegrid/clb` fuzzer. Here is the whole trace:

```
Traceback (most recent call last):
File "add_tdb.py", line 162, in <module>
     main()
   File "add_tdb.py", line 158, in main
     run(args.fn_in, args.fn_out, verbose=args.verbose)
   File "add_tdb.py", line 138, in run
     for (tile, frame, wordidx) in load_db(tdb_fn):
   File "add_tdb.py", line 51, in load_db
     check_frames(tagstr, addrlist)
  File "add_tdb.py", line 21, in check_frames
     frame = parse_addr(addrstr, get_base_frame=True)
   File "add_tdb.py", line 30, in parse_addr
     frame = int(line[0], 16)
 ValueError: invalid literal for int() with base 16: '<0'
```

Since all the DB builds fail with the same error, I suspect the changes in the tools trigger this. @LAK132 can you check this?

LAK132 committed:
Remove changes to fasm2frames.py

Signed-off-by: Lakira Ashley <lakridesagain@gmail.com>

mithro commented:
@LAK132 - If you are having issues getting the tool / testing environment setup, please log github issues and we will try and improve both the instructions and the process.

hansfbaier commented:
![image](https://user-images.githubusercontent.com/148607/214445108-d97b367b-2ffb-44f4-a0ca-79c2acdea510.png)
segbits_tilegrid.tdb for xc7a100t has no candidates entries

hansfbaier commented:
I narrowed down the problem to the `bitread` tool: When I exchange it with a known good version then the candidates are found.

hansfbaier commented:
@LAK132 This change in configuration.h looks suspicious to me, it possibly has different semantics than the original code:
![image](https://user-images.githubusercontent.com/148607/214451944-0b174f6b-017e-43ec-b332-733491f701d2.png)

hansfbaier commented:
@LAK132 Here are two example bitstreams which show the difference:
[example-data.tar.gz](https://github.com/f4pga/prjxray/files/10495058/example-data.tar.gz)
Consider the .bits files there (the output of bitread).
the -bad version (with the bitread tool from this branch) seems to have extraneous data:
![image](https://user-images.githubusercontent.com/148607/214453394-a27548f9-b95a-45fe-9ccf-5fd330536fa9.png)

LAK132 commented:
> it possibly has different semantics than the original code

Technically true. Previously if that address had already been written to it would overwrite it, whereas with `.insert` it will not overwrite existing frame writes. That said, writing to the same frame multiple times feels like a bug? The beauty of `.insert` is that it won't default construct a frame before the assignment can take place. That said, that change should have been the only major change to the uncompressed codepath, so if reverting that fixes the problem...

LAK132 commented:
Also it looks like with that example data, it's the -good output that has the extra data, not the -bad output? Were these outputs made with or without compression enabled?

hansfbaier commented:
@LAK132 Sorry, I was wrong, the original bitread has the extra bits (green parts above), the bitread of this branch does not have them.

hansfbaier commented:
@LAK132 Here is a 'good' version of bitread for comparison
[bitread.gz](https://github.com/f4pga/prjxray/files/10495654/bitread.gz)

hansfbaier commented:
@LAK132 So the way to go would be:
1. instrument the code with printfs to find out what happenes with the missing parts, or use a debugger
2. run `./build/tools/bitread -part_file database/artix7/xc7a100tfgg676-1/part.yaml -F 0x00000000:0xffffffff -o design.bits -y -z design.bit` (take design.bit from the example data, either the good or the bad version, they are equivalent)

LAK132 committed:
Use operator[] instead of .insert

Signed-off-by: Lakira Ashley <lakridesagain@gmail.com>

LAK132 commented:
`.insert_or_assign` would have been a better way to do this but unfortunately that's a C++17 feature

hansfbaier commented:
@LAK132 Did that fix the bitread results?

LAK132 commented:
It should have, the uncompressed codepath should be practically untouched now

hansfbaier commented:
Can you test it? End of the day for me here...

LAK132 commented:
The bitread output appears to be the same as the known good version with my test core project. I still haven't managed to run any of the other tests myself.

hansfbaier commented:
@LAK132 OK great let's wait for the CI then

hansfbaier commented:
@LAK132 Good news, the changes work well here on my machine. I added a temporary PR (see above) to run the full CI on this build. @kgugala @mithro Can we run the CI on this?

hansfbaier commented:
@LAK132 CI results look great! (Only the gtp-common-pips fuzzer is red, but that one is notoriously flickering) @mkurc-ant @kgugala Ready for review!

hansfbaier commented:
@tmichalak Would it be possible to run the CI on this branch?

LAK132 committed:
Remove C++14 directive (fix conflict with PR #2118)

Signed-off-by: Lakira Ashley <lakridesagain@gmail.com>

LAK132 commented:
Is there a reason the tests were cancelled? Did they just time out?

LAK132 commented:
All green now, anything else I need to do?

LAK132 commented:
Any update on this?

LAK132 committed:
Merge remote-tracking branch 'upstream/master' into HEAD

LAK132 committed:
Fix missing <cstdint> include

Signed-off-by: Lakira Ashley <lakridesagain@gmail.com>
